README.md
# 封装 services 请求

## 加入了 husky 代码提交规范的校验

## 加入了 standard-version 自动生成 Changelog 日志和自动发版的指令

```json
  "devDependencies": {
        "standard-version": "^9.5.0"
  }


  "scripts": {
    "build": "rollup -c",
    "release": "standard-version && npm publish",
    "prepare": "husky install"
  },

```

```js
import { HttpRequest } from "@lg/lg-services";
import { ElMessage } from "element-plus";
const options = {
  baseURL: "",
  withCredentials: false,
  timeout: 30 * 1000,
  headers: {
    "Access-Control-Allow-Origin": "*",
  },
  successCode: 0, //业务状态码-即服务端业务请求正确的状态码(区分与浏览器的ok的状态码,浏览器正常的状态是200,)
  messageCallback: (message, type: any = "error") => {
    ElMessage({
      message,
      type,
    });
  },
};

const { instance: ajax } = new HttpRequest(options);
export { ajax };
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.16](https://github.com/leigong421/lg-services/compare/v0.0.15...v0.0.16) (2023-05-13)

### [0.0.15](https://github.com/leigong421/lg-services/compare/v0.0.14...v0.0.15) (2023-05-13)

### [0.0.14](https://github.com/leigong421/lg-services/compare/v0.0.13...v0.0.14) (2023-05-13)

### [0.0.13](https://github.com/leigong421/lg-services/compare/v0.0.12...v0.0.13) (2023-05-13)

### [0.0.12](https://github.com/leigong421/lg-services/compare/v0.0.11...v0.0.12) (2023-05-13)

### [0.0.11](https://github.com/leigong421/lg-services/compare/v0.0.10...v0.0.11) (2023-05-13)

### [0.0.10](https://github.com/leigong421/lg-services/compare/v0.0.9...v0.0.10) (2023-05-13)

### [0.0.9](https://github.com/leigong421/lg-services/compare/v0.0.8...v0.0.9) (2023-05-13)

### [0.0.8](https://github.com/leigong421/lg-services/compare/v0.0.7...v0.0.8) (2023-05-13)

### [0.0.7](https://github.com/leigong421/lg-services/compare/v0.0.6...v0.0.7) (2023-05-13)

### [0.0.6](https://github.com/leigong421/lg-services/compare/v0.0.1...v0.0.6) (2023-05-13)


### Docs

* 增加功能说明文档 ([91620a5](https://github.com/leigong421/lg-services/commit/91620a58ee800344001c5affdf51f7786145e391))


### Others

* **release:** 0.0.3 ([cf4d3c1](https://github.com/leigong421/lg-services/commit/cf4d3c18c612327ed4b8dc4aa14eae023a59aae2))
* **release:** 0.0.4 ([fbcc4c1](https://github.com/leigong421/lg-services/commit/fbcc4c18a7f0dd923eec187ebfc278bf2f82bb5b))
* **release:** 0.0.5 ([470234b](https://github.com/leigong421/lg-services/commit/470234b947a0581bf73a3ccce3fad4a44086db00))

### [0.0.5](https://github.com/leigong421/lg-services/compare/v0.0.1...v0.0.5) (2023-05-13)


### Others

* **release:** 0.0.3 ([cf4d3c1](https://github.com/leigong421/lg-services/commit/cf4d3c18c612327ed4b8dc4aa14eae023a59aae2))
* **release:** 0.0.4 ([fbcc4c1](https://github.com/leigong421/lg-services/commit/fbcc4c18a7f0dd923eec187ebfc278bf2f82bb5b))


### Docs

* 增加功能说明文档 ([91620a5](https://github.com/leigong421/lg-services/commit/91620a58ee800344001c5affdf51f7786145e391))

### [0.0.4](https://github.com/leigong421/lg-services/compare/v0.0.1...v0.0.4) (2023-05-13)


### Others

* **release:** 0.0.3 ([cf4d3c1](https://github.com/leigong421/lg-services/commit/cf4d3c18c612327ed4b8dc4aa14eae023a59aae2))

### [0.0.3](https://github.com/leigong421/lg-services/compare/v0.0.1...v0.0.3) (2023-05-13)

### [0.0.1](https://github.com/leigong421/lg-services/compare/v0.0.5...v0.0.1) (2023-05-13)


### Others

* **release:** 0.0.2 ([ea2c6af](https://github.com/leigong421/lg-services/commit/ea2c6af4963f95890f302d0ee446048a2a956014))
* **release:** 0.0.3 ([80b9bb6](https://github.com/leigong421/lg-services/commit/80b9bb6552eb97935250692ac3b919f32eafcbb8))
* **release:** 0.0.4 ([6009e3a](https://github.com/leigong421/lg-services/commit/6009e3a714ee717f77c0476c619df1ecd20653c6))

### [0.0.2](https://github.com/leigong421/lg-services/compare/v0.0.5...v0.0.2) (2023-05-13)


### Others

* **release:** 0.0.3 ([80b9bb6](https://github.com/leigong421/lg-services/commit/80b9bb6552eb97935250692ac3b919f32eafcbb8))
* **release:** 0.0.4 ([6009e3a](https://github.com/leigong421/lg-services/commit/6009e3a714ee717f77c0476c619df1ecd20653c6))

### [0.0.4](https://github.com/leigong421/lg-services/compare/v0.0.5...v0.0.4) (2023-05-13)


### Others

* **release:** 0.0.3 ([80b9bb6](https://github.com/leigong421/lg-services/commit/80b9bb6552eb97935250692ac3b919f32eafcbb8))


__tests__/functions.js
/*
 * @Author: suchiva@126.com
 * @Date: 2022-09-15 15:52:21
 * @LastEditors: your Name
 * @LastEditTime: 2022-09-15 16:10:26
 * @Description: 测试文件
 */
import { array } from './base';


.commitlintrc.js
module.exports = {
  // 继承的规则
  extends: ['@commitlint/config-conventional'],
  // 定义规则类型
  rules: {
    // type 类型定义，表示 git 提交的 type 必须在以下类型范围内
    'type-enum': [
      2,
      'always',
      [
        'feat', // 增加新功能
        'fix', // 修复 bug
        'add', // 增加代码逻辑
        'del', // 删除功能
        'update', // 更新功能
        'docs', // 文档相关的改动
        'style', // 不影响代码逻辑的改动，例如修改空格，缩进等
        'build', // 构造工具或者相关依赖的改动
        'refactor', //  代码重构
        'revert', // 撤销，版本回退
        'test', // 添加或修改测试
        'perf', // 提高性能的改动
        'chore', // 修改 src 或者 test 的其余修改，例如构建过程或辅助工具的变动
        'ci' // CI 配置，脚本文件等改动
      ]
    ],
    // subject 大小写不做校验
    'subject-case': [0]
  }
  // plugins: [
  //   {
  //     rules: {
  //       "commit-rule": ({ raw }) => {
  //         console.log(8888888);
  //         return [
  //           /^\[(feat|fix|add|del|update|docs|style|build|refactor|revert|test|perf|chore)].+/g.test(
  //             raw
  //           ),
  //           `commit备注信息格式错误，格式为 <[type] 修改内容>，type支持${types.join(
  //             ","
  //           )}`,
  //         ];
  //       },
  //     },
  //   },
  // ],
};


src/ajax/index.js
import axios from "axios";

const MESSAGE_NOTICE = {
  503: "服务器内部异常!",
  401: "访问凭证无效，请重新登录授权!",
  403: "访问凭证无效，请重新登录授权!",
};

class HttpRequest {
  constructor(options) {
    this.instance = axios.create(options);
    // this.instance.CancelToken = axios.CancelToken;
    // 手动挂载2个方法
    this.instance.all = axios.all;
    this.instance.spread = axios.spread;
    this.instance.token = options?.customParams?.token;
    this.instance.setToken = (token) => {
      this.instance.token = token;
    };
    this.addRequestInterceptor(options);
    this.addResponseInterceptor(options);
  }
  //请求拦截
  addRequestInterceptor(options) {
    this.instance.interceptors.request.use(
      (config) => {
        // const token =
        //   this.instance.token ||
        //   JSON.parse(window.sessionStorage.getItem('vuex'))?.user?.token;
        // config.headers.token = token || '';

        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );
  }

  //相应拦截
  addResponseInterceptor(options) {
    this.instance.interceptors.response.use(
      (response) => {
        // console.log('interceptors.response', options);
        //返回数据是json格式处理函数
        function jsonHandler(response) {
          const { code, msg } = response.data;
          const res = response.data;
          if (code === options.successCode) {
            return response.data;
          } else {
            options.messageCallback(res.msg || "Error", options?.componentType);
            return Promise.reject(res);
          }
        }

        if (response.status === 200) {
          //返回的数据不是对象
          if (
            Object.prototype.toString.call(response.data) !== "[object Object]"
          ) {
            return response;
          }
          return jsonHandler(response);
        } else {
          options.messageCallback(
            response.statusText || "Error",
            options?.componentType
          );
          return Promise.reject(response);
        }
      },
      (error) => {
        options.messageCallback(error, options?.componentType);
        return Promise.reject(error);
      }
    );
  }
}

export default HttpRequest;


rollup.config.js
/*
 * @Author: lg
 * @Date: 2022-08-30 15:49:48
 * @LastEditors: lg
 * @LastEditTime: 2023-05-13 13:51:10
 * @Description: 打包
 */
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import babel from 'rollup-plugin-babel';
import { terser } from 'rollup-plugin-terser';
import cleanup from 'rollup-plugin-cleanup';
// : {
//   file: './dist/index.js',
//   format: 'umd', // umd是兼容amd/cjs/iife的通用打包格式，适合浏览器
//   name: 'bundleName',
//   globals: {
//     axios: 'axios'
//   }
// }
export default {
  input: './src/index.js',
  output: [
    {
      file: './dist/index.umd.js',
      name: 'index',
      format: 'umd'
    },
    {
      file: './dist/index.es.js',
      format: 'es'
    }
  ],

  plugins: [
    babel({
      presets: ['@babel/preset-env'],
      exclude: 'node_modules/**', // 防止打包node_modules下的文件
      runtimeHelpers: true // 使plugin-transform-runtime生效
    }),
    resolve(),
    commonjs(),
    terser({
      compress: {
        pure_funcs: ['console.log'] // 去掉console.log函数,压缩js
      }
    }),
    cleanup() //清除注释
  ],
  external: ['axios']
};
